### ISSUE TITLE
implement enum

### ISSUE THREAD

**Issue by jameysharp** [jameysharp/corrode#19]
C's `enum` types are pretty easy to translate except for the fact that C allows arbitrary integer values to be used where an enum variant is expected. Rust's enums don't allow that (as far as I know).

So step one is: Figure out what we can translate C enums to in Rust that will allow valid C programs to be translated (even if they use values outside the range of the enum), while ideally preserving all uses of the names of the enum variants.

**Comment by dgavedissian** [comment]
Turns out that Rust's enums do allow this style if you add an explicit typecast when using them as integer values: http://stackoverflow.com/a/26949013

I just double checked and this compiles with Rust 1.9.

**Comment by jameysharp** [comment]
Good find, @davedissian, and thanks for the pointer! I'm pretty sure we should use that as part of the Rust translation of enums.

However, my concern was for code like this:

``` c
enum E { A, B, C } x = 42;
```

This is legal C; an enum-typed variable can hold any int value, even if there's no enum label that corresponds to that value.

A naive Rust translation doesn't work. This program fails to compile with error "(expected enum `E`, found integral variable)":

``` rust
#[derive(Debug)]
pub enum E { A, B, C }
fn main() {
    let x : E = 42;
    println!("{:?}", x);
}
```

Even if the integer value corresponds to one of the enum labels, we may not be able to prove it at compile time. Here's a legal C function:

``` c
enum E f(int i) { return i; }
```

I suspect what this means is that Corrode needs to translate any enum type names to `i32`, although we could at least emit a type alias, so that function translates to:

``` rust
type E = i32;
fn f(i: i32) -> E { i }
```

The next question is, can we at least preserve the names of the enum labels? It'd be nice to translate this:

``` c
enum L { A = 1, B = 27 };
enum L v = B;
```

to something like this:

``` rust
enum _L { A = 1, B = 27 }
type L = i32;
let v : L = _L::B as i32;
```

so that we don't lose the fact that it was supposed to be the `B` variant of this enum.

Unfortunately, C does not require enum variants to have different values, but [Rust does](https://doc.rust-lang.org/error-index.html#E0081). Here's an example from [musl](https://www.musl-libc.org/), in `include/arpa/nameser.h`:

``` c
    enum __ns_sect {
        ns_s_qd = 0,
        ns_s_zn = 0,
        ns_s_an = 1,
        ns_s_pr = 1,
        ns_s_ns = 2,
        ns_s_ud = 2,
        ns_s_ar = 3,
        ns_s_max = 4
    }
```

So we might need to translate the enum labels to Rust `const` items or something instead, which means we lose the information that they were all supposed to go together as one type.

If there are better solutions to any of these problems, I'd love to hear about them!

**Comment by joshtriplett** [comment]
@jameysharp Is storing arbitrary non-enum values in an enum allowed by the spec, or unspecified/unknown behavior that compilers tend to allow (and optionally warn about)?

Quoting C99:

>  Each enumerated type shall be compatible with char, a signed integer type, or an unsigned integer type. The choice of type is implementation-defined, but shall be capable of representing the values of all the members of the enumeration.

**Comment by jameysharp** [comment]
Hi @joshtriplett! I think that all the spec is saying there is that the choice of which integer type a particular enum is compatible with is implementation-defined, subject to the constraint that all the enumerator values must fit in whatever type the implementation chose.

Looks like the simplest answer to your question begins with C99 section 6.2.5 (Language/Concepts/Types). Paragraphs 17 and 18 say,

> The type `char`, the signed and unsigned integer types, and the enumerated types are
> collectively called _integer types_. ...
> 
> Integer and floating types are collectively called _arithmetic types_. ...

Then in section 6.5.16.1 "Simple assignment", we have:

> One of the following shall hold:
> — the left operand has qualified or unqualified arithmetic type and the right has arithmetic type;
> ...
> In _simple assignment_ (`=`), the value of the right operand is converted to the type of the assignment expression and replaces the value stored in the object designated by the left operand.

Together, I think this means that when the left-hand side is an enumerated type, the right-hand side can be any arithmetic type. (Including floating-point or complex types, bleh!) And what does the conversion actually do? That's in section 6.3.1.3:

> When a value with integer type is converted to another integer type other than `_Bool`, if the value can be represented by the new type, it is unchanged.

Followed by the usual rules for unsigned integer wrap and implementation-defined signed wrap behavior.

In short: my reading of the spec is that yes, this wackiness is specifically allowed. Boo. :frowning:

**Comment by joshtriplett** [comment]
>  In simple assignment (=), the value of the right operand is converted to the type of the assignment expression and replaces the value stored in the object designated by the left operand.
> 
> When a value with integer type is converted to another integer type other than _Bool, if the value can be represented by the new type, it is unchanged.

If you're willing to be different than all the compilers out there, can "can be represented by the new type" be read to exclude values other than those of the enum?  Could you say that `enum E { A, B };` can't represent `10`?  Or does the definition of "can be represented" and of the enum type exclude that interpretation?

**Comment by jameysharp** [comment]
Good effort, as I too would very much prefer your interpretation, but I'm _pretty_ sure it would not be conforming to the C standard. :disappointed: I'm also pretty sure there's a lot of C code out there which would break with that interpretation.

OK, I'm not satisfied with that answer, let's try a more formal one. As far as I can tell, the "_compatible type_" relation is supposed to be symmetric: it's defined in C99 6.2.7 and the language there seems to just assume so. The section you quoted earlier says, "Each enumerated type shall be compatible with" an implementation-defined integer type. And section 6.3 "Conversions", paragraph 2, says:

> Conversion of an operand value to a compatible type causes no change to the value or the
> representation.

So I read that as not only can you convert any value of enumerated type to its compatible integer type, but you can also convert any value of the integer type to the enumerated type.

**Comment by joshtriplett** [comment]
Oh well.

This seems like the kind of thing that could potentially become an option in the future ("strict enums"), as many C programs will work fine with that interpretation.  (Far more programs declare variables of type `int` or similar and store enum values in them, or make assumptions about enum sizes, but I think relatively few C programs assume they can store arbitrary non-enum values in an enum.)

**Comment by adnelson** [comment]
@jameysharp in the readme for this repo, you state

> If this program can't translate your C source to equivalent Rust, you might consider whether your program is too complicated and hiding bugs.

The way I read that, it implicitly means that there's no guarantee that corrode will translate _any_ valid C program to Rust. This seems like a prime candidate for this case. Wouldn't it be conceivable to simply reject the "enums-as-integers" duality and force programs to use only values defined in the enum? After all, if the integer being assigned _is_ a valid member of the enum, modifying the code to satisfy corrode should be easy in most cases.

**Comment by Storyyeller** [comment]
Why not just choose an integral type (e.g. i32) and convert all enums to that? Seems like that is both simple and maximizes compatibility.

**Comment by jameysharp** [comment]
OK, between @adnelson and @Storyyeller we have two strong proposals for how to handle enum types.
- I like @adnelson's proposal because it preserves as much semantic information from the source program as possible, at the cost of rejecting some valid C programs.
- I like @Storyyeller's proposal because whenever there's a reasonable way we can accept a random C program, I'd like to take it.

So I think the way to go is:
1. Implement @adnelson's proposal, for the reasons given&mdash;it's useful for finding bugs. As a first approximation, report a translation-unimplemented error any time a value of a different type is assigned to a variable of this enum type.
2. Even with strict checking, we can allow assigning numeric literals to enum-typed variables if the numeric constant matches the value of one of the type's declared values; we can translate it to the appropriate name in the generated Rust source. This is a nice follow-up task.
3. Eventually provide some mechanism where the programmer can opt out of the strict checking and switch to @Storyyeller's translation, on a per-type basis. I think the way to do that is to allow a custom attribute on enum type declarations which indicates that this enum type must be able to hold arbitrary `int`s. We could also automatically trigger this mode if the enum has multiple names with the same value, like the `__ns_sect` type from musl, which I quoted above.

**Comment by jameysharp** [comment]
I should have noted that this is basically what @joshtriplett proposed, just fleshed out with more detail. :smile:

**Comment by jameysharp** [comment]
I'm not sure why I didn't close this issue before, since I merged the pull request implementing `enum` support some time ago.
